src/intrinsics/ecma262/Object.js

```diff
@@ -41,6 +41,92 @@ import { Create, Havoc, Properties as Props, To } from "../../singletons.js";
+//
+// This is a recursive function, so it will attempt to do this multiple times
+// until it can no longer do so, or if a particular Object.assign is visited
+// and thus it is not possible to merge the Object.assign calls together.
+export function attemptToMergeEquivalentObjectAssigns(callbacks: VisitEntryCallbacks, value: Value): boolean {
+  if (!(value instanceof AbstractObjectValue)) {
+    return false;
+  }
+  let realm = value.$Realm;
+  let temporalBuildNodeEntry = realm.getTemporalBuildNodeEntryArgsFromDerivedValue(value);
+  if (temporalBuildNodeEntry === undefined) {
+    return false;
+  }
+  let args = temporalBuildNodeEntry.args;
+  // If we are Object.assigning 3 or more args
+  if (args.length < 3) {
+    return false;
+  }
+  // Then scan through the args after the "to" of this Object.assign, to see if any
+  // other sources are the "to" of a previous Object.assign call
+  for (let i = 2; i < args.length; i++) {
+    let possibleOtherObjectAssignTo = args[i];
+    // Ensure that the "to" value can be omitted
+    if (!callbacks.canOmit(possibleOtherObjectAssignTo)) {
+      continue;
+    }
+    // Check if the "to" was definitely an Object.assign, it should
+    // be a snapshot AbstractObjectValue
+    if (
+      possibleOtherObjectAssignTo instanceof AbstractObjectValue &&
+      // We don't handle conditions for now, this can be done at a later stage
+      possibleOtherObjectAssignTo.kind !== "conditional"
+    ) {
+      let otherTemporalBuildNodeEntry = realm.getTemporalBuildNodeEntryArgsFromDerivedValue(
+        possibleOtherObjectAssignTo
+      );
+      if (otherTemporalBuildNodeEntry === undefined) {
+        continue;
+      }
+      let otherArgs = otherTemporalBuildNodeEntry.args;
+      let objectAssingFunc = realm.intrinsics.Object.$Get("assign", realm.intrinsics.Object);
+      // Object.assign has at least 2 args and the first is that of Object.assign native function
+      if (otherArgs.length < 3 || otherArgs[0] !== objectAssingFunc) {
+        continue;
+      }
+      let to = args[1];
+      // If we cannot omit the "to" value that means it's being used, so we shall not try to
+      // optimize this Object.assign.
+      if (!callbacks.canOmit(to)) {
+        let newArgs = [objectAssingFunc, to];
+        for (let x = 2; x < otherArgs.length; x++) {
+          newArgs.push(otherArgs[x]);
+        }
+        for (let x = 2; x < args.length; x++) {
+          let arg = args[x];
+          // We don't want to add the "to" that we're merging with!
+          if (arg !== possibleOtherObjectAssignTo) {
+            newArgs.push(arg);
+          }
+        }
+        // We now mutate the args in place with the above new args, clearing out the old args.
+        // The end result should be a merged Object.assign and one less temporal entry.
+        // The previous Object.assign temporal's "to" is no longer being used anywhere and should
+        // now dead-code away nicely.
+        args.length = 0;
```

Comment:
In my view, all in-place mutations of `GeneratorEntry` instances are undesirable. Yes, we already have a few such instances, but I'd rather want to make those away than to pile-up more. (The background is that this makes the visitor terribly stateful, and running it twice could yield different results, which is just wrong.) What I'd rather want to do is to maintain some information on the side (possibly managed by the visitor and exposed via the `callbacks`, but nevertheless separate information).